cpp/src/groupby/hash/groupby_kernels.cuh

```diff
@@ -0,0 +1,296 @@
+
+    cudf::genericAtomicOperation(&target_data[target_index], TargetType{1},
+                                 DeviceSum{});
+  }
+};
+
+template <bool source_has_nulls>
+struct elementwise_aggregator {
+  template <typename SourceType>
+  __device__ inline void operator()(gdf_column const& target,
+                                    gdf_size_type target_index,
+                                    gdf_column const& source,
+                                    gdf_size_type source_index, operators op) {
+    switch (op) {
+      case MIN: {
+        update_target_element<SourceType, MIN, source_has_nulls>{}(
+            target, target_index, source, source_index);
+        break;
+      }
+      case MAX: {
+        update_target_element<SourceType, MAX, source_has_nulls>{}(
+            target, target_index, source, source_index);
+        break;
+      }
+      case SUM: {
+        update_target_element<SourceType, SUM, source_has_nulls>{}(
+            target, target_index, source, source_index);
+        break;
+      }
+      case COUNT: {
+        update_target_element<SourceType, COUNT, source_has_nulls>{}(
+            target, target_index, source, source_index);
+      }
+      default:
+        return;
+    }
+  }
+};
+
+/**---------------------------------------------------------------------------*
+ * @brief Performs an in-place update by performing elementwise aggregation
+ * operations between a target and source row.
+ *
+ * For `i` in `[0, num_columns)`, each element in the target row is updated as:
+ *
+ *```
+ * target_row[i] = target_row[i] op[i] source_row[i]
+ *```
+ * @note If a source element is NULL, the aggregation operation for
+ * that column is skipped.
+ *
+ * @note If a target element is NULL, it is assumed that the value of the NULL
+ * element is the identity value of the aggregation operation being performed.
+ * The aggregation operation is performed between the source element and the
+ * identity value, and the target element's bit is set to indicate it is no
+ * longer NULL.
+ *
+ * @note For COUNT, it is assumed the target element can *never* be NULL. As
+ * such, it is expected the target element's bit is already set.
+ *
+ * @param target Table containing the target row
+ * @param target_index Index of the target row
+ * @param source Table cotaning the source row
+ * @param source_index Index of the source row
+ * @param ops Array of operators to perform between the elements of the
+ * target and source rows
+ *---------------------------------------------------------------------------**/
+template <bool values_have_nulls = true>
+__device__ inline void aggregate_row(device_table const& target,
+                                     gdf_size_type target_index,
+                                     device_table const& source,
+                                     gdf_size_type source_index,
+                                     operators* ops) {
+  using namespace bit_mask;
+  thrust::for_each(
```

Comment:
What's the benefit of using a sequential `thrust::for_each` with a counting iterator here?  A for loop would be much clearer.  `for (gdf_index_type i = 0; i < target.num_columns(); ++i) {...}` and there would be no need for a lambda...

You could maybe do `for (auto col : target) {}` but since you need both the source and target columns it might be a faff.